server/src/main/scala/org/http4s/server/middleware/CSRFMiddleware.scala

```diff
@@ -0,0 +1,196 @@
+          c2 <- OptionT(
+            Task.point(r.headers.get(CaseInsensitiveString(config.headerName))))
+          raw1 <- extractRaw(c1.content)
+          raw2 <- extractRaw(c2.value)
+          response <- if (CSRFMiddleware.isEqual(raw1, raw2)) {
+            OptionT[Task, MaybeResponse](service(r).map(Some(_)))
+          } else {
+            OptionT.none[Task, MaybeResponse]
+          }
+          newToken <- OptionT[Task, String](signToken(raw1).map(Some(_)))
+        } yield
+          response.cata(
+            _.addCookie(Cookie(config.cookieName, newToken)),
+            Response(Status.NotFound)))
+          .getOrElse(Response(Status.Unauthorized))
+      }
+  }
+
+  /** Embed a token into a response **/
+  def embedNew(res: MaybeResponse): Task[Response] =
+    generateToken.map(
+      content =>
+        res.cata(_.addCookie(Cookie(config.cookieName, content)),
+          Response(Status.NotFound)))
+
+  /** Middleware to embed a csrf token into routes that do not have one. **/
+  def withNewToken: Middleware[Request, MaybeResponse, Request, MaybeResponse] =
+    _.andThen(Kleisli(embedNew))
+
+}
+
+object CSRFMiddleware {
+
+  /** An instance of SecureRandom to generate
+    * tokens, properly seeded:
+    * https://tersesystems.com/blog/2015/12/17/the-right-way-to-use-securerandom/
+    */
+  private lazy val CachedRandom = {
+    val r = new SecureRandom()
+    r.nextBytes(new Array[Byte](20))
+    r
+  }
+
+  val SigningAlgo = "HmacSHA1"
+  val SHA1ByteLen = 20
+  val CSRFTokenLength = 32
+
+  /** Hex encoding digits. Adapted from apache commons Hex.encodeHex **/
+  private val Digits: Array[Char] = Array('0', '1', '2', '3', '4', '5', '6',
+    '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
+
+  protected def find[A](nel: NonEmptyList[A], p: A => Boolean): Option[A] =
+    if (p(nel.head)) {
+      Some(nel.head)
+    } else {
+      nel.tail.find(p)
+    }
+
+  protected def cookieFromHeaders(request: Request,
+    headerName: String): OptionT[Task, Cookie] =
+    OptionT(
+      Task.point(HCookie
+        .from(request.headers)
+        .flatMap(v => CSRFMiddleware.find[Cookie](v.values, _.name == headerName))))
+
+  /** A Constant-time string equality **/
+  def isEqual(s1: String, s2: String): Boolean =
+    MessageDigest.isEqual(s1.getBytes(StandardCharsets.UTF_8),
+      s2.getBytes(StandardCharsets.UTF_8))
+
+  /** Encode a string to a Hexadecimal string representation
+    * Adapted from apache commons Hex.encodeHex
+    */
+  protected def encodeHex(data: Array[Byte]): Array[Char] = {
+    val l = data.length
+    val out = new Array[Char](l << 1)
+    // two characters form the hex value.
+    var i = 0
+    var j = 0
+    while (i < l) {
+      out(j) = Digits((0xF0 & data(i)) >>> 4)
+      j += 1
+      out(j) = Digits(0x0F & data(i))
+      j += 1
+      i += 1
+    }
+    out
+  }
+
+  /** Generate an unsigned CSRF token from a `SecureRandom` **/
+  protected def genTokenString: String = {
```

Comment:
Why protected in an object?

This looks effectful, but if it's not public API and unwrapped for efficiency, that's okay.

Reply:
oh my bad, didn't see this one. Yeah, i'm simply not surrounding it in `Task` for efficiency. I can make private.